File: src/litcoder/module-3/lab-2/LongestSubstringContainingVowel.py
from typing import Dict

def solution(string: str) -> int:
    """Returns the length of longest substring of the passed string which
    contains even number of vowels

    Args:
        string (str): string

    Returns:
        int: length of longest substring with even number of vowels
    """
    longest_substring_length = 0

    # iterate over all the substring of given string
    for i in range(len(string)):
        for j in range(i, len(string)):
            # get count vowels in current substring
            vowel_count = count_vowels(string[i : j + 1])

            # check all count is even
            all_even = True
            for count in vowel_count.values():
                if not is_even(count):
                    all_even = False
                    break
            if not all_even:
                continue

            # replace length of longest_substring_length
            longest_substring_length = max(longest_substring_length, j + 1 - i)

    return longest_substring_length


def is_even(n: int) -> bool:
    return (n & 1) == 0


def count_vowels(string: str) -> Dict[str, int]:
    """Returns the counts of each vowel in the string as dictionary

    Args:
        string (str): string to count vowels in

    Returns:
        dict[str, int]: dictionary mapped with vowel as key and its count its value
    """
    vowel_counts = {"a": 0, "e": 0, "i": 0, "o": 0, "u": 0}
    for char in string:
        if char in vowel_counts.keys():
            vowel_counts[char] += 1

    return vowel_counts


testcases: int = int(input())
for _ in range(testcases):
    string = input()
    print(solution(string))

File: src/codes/college/litcoder/module3/lab2/LongestSubstringContainingVowel.java
package codes.college.litcoder.module3.lab2;

import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class LongestSubstringContainingVowel {

    public static int solution(String string) {
        int longestSubstringLength = 0;

        // iterate over all substrings of the given string
        for (int i = 0; i < string.length(); i++) {
            for (int j = i; j < string.length(); j++) {
                // get count of vowels in the current substring
                Map<Character, Integer> vowelCount = countVowels(
                        string.substring(i, j + 1));

                // check if all counts are even
                boolean allEven = true;
                for (int count : vowelCount.values()) {
                    if (!isEven(count)) {
                        allEven = false;
                        break;
                    }
                }
                if (!allEven) {
                    continue;
                }

                // update length of longestSubstringLength
                longestSubstringLength = Math.max(longestSubstringLength, j + 1 - i);
            }
        }

        return longestSubstringLength;
    }

    public static boolean isEven(int n) {
        return (n & 1) == 0;
    }

    public static Map<Character, Integer> countVowels(String substring) {
        Map<Character, Integer> vowelCounts = new HashMap<>();
        vowelCounts.put('a', 0);
        vowelCounts.put('e', 0);
        vowelCounts.put('i', 0);
        vowelCounts.put('o', 0);
        vowelCounts.put('u', 0);

        for (char ch : substring.toCharArray()) {
            if (vowelCounts.containsKey(ch)) {
                vowelCounts.put(ch, vowelCounts.get(ch) + 1);
            }
        }

        return vowelCounts;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int testcases = scanner.nextInt();
        scanner.nextLine(); // consume the newline after reading the integer

        for (int t = 0; t < testcases; t++) {
            String string = scanner.nextLine();
            System.out.println(solution(string));
        }
    }
}
